File: src/main/java/SmartCasts.java
package main.java;

class Shape {

}

class Square extends Shape{
    public Square(int side) {
        this.side = side;
    }

    public int side;
}


class Rectangle extends Shape{
    public Rectangle(int length, int width) {
        this.length = length;
        this.width = width;
    }

    public int length;
    public int width;

}

class PerimeterCalculator {
    public static double perimeter(Shape shape){
        if(shape instanceof Rectangle){
            Rectangle rectangle = (Rectangle) shape;
            return 2 * rectangle.length * rectangle.width;
        }
        else if (shape instanceof Square){
            Square square = (Square) shape;
            return 4 * square.side;
        }
        return 0;
    }
}

public class SmartCasts {

    public static void main(String[] args) {

        Square square = new Square(5);
        Rectangle rectangle = new Rectangle(4 ,2);

        System.out.println(PerimeterCalculator.perimeter(square));
        System.out.println(PerimeterCalculator.perimeter(rectangle));

    }

}

File: src/main/kotlin/SmartCasts.kt
open class Shape() {}

data class Square(val side: Int) : Shape()
data class Rectangle(val length: Int, val width: Int): Shape()

class PerimeterCalculator {
    fun perimeter(shape: Shape): Int {
        return when(shape) {
            is Square ->
                4 * shape.side
            is Rectangle ->
                2 * (shape.length + shape.width)
            else ->
                0
        }
    }
}

fun main() {
    val square = Square(4)
    val rectangle = Rectangle(4, 2)

    val perimeterCalculator = PerimeterCalculator()
    println(perimeterCalculator.perimeter(square))
    println(perimeterCalculator.perimeter(rectangle))
}
